### ISSUE TITLE
attempt to support "go build -buildmode=plugin -linkshared",it build success but run failed now.

### ISSUE THREAD

**Issue by crazycomputer** [golang/go#45692]
### What version of Go are you using (`go version`)?

<pre>
$ go version
go version go1.14.12 linux/amd64
</pre>

### Does this issue reproduce with the latest release?
yes

### What operating system and processor architecture are you using (`go env`)?

<details><summary><code>go env</code> Output</summary><br><pre>
$ go env
GO111MODULE=""
GOARCH="amd64"
GOBIN=""
GOCACHE="/root/.cache/go-build"
GOENV="/root/.config/go/env"
GOEXE=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOINSECURE=""
GONOPROXY=""
GONOSUMDB=""
GOOS="linux"
GOPATH="/root/go"
GOPRIVATE=""
GOPROXY="https://proxy.golang.org,direct"
GOROOT="/mnt/d/wsl/01.Tools/golang/go"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/mnt/d/wsl/01.Tools/golang/go/pkg/tool/linux_amd64"
GCCGO="gccgo"
AR="ar"
CC="gcc"
CXX="g++"
CGO_ENABLED="1"
GOMOD=""
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build174216265=/tmp/go-build -gno-record-gcc-switches"

</pre></details>

### What did you do?
first, use  `go install -buildmode=shared -linkshared std ` to build and install libstd.so
second, use `go build -buildmode=plugin -linkshared plugin.go` to build plugin.so
plugin.go:
```
package main
import "fmt"
func Print(){
	fmt.Println("plugin")
}
```
then, use `go build -linkshared main.go` to build main.go whit linkshared
main.go
```
package main

import (
	"log"
	"plugin"
)

func main() {
	p, err := plugin.Open("./plugin.so")
	if err != nil {
		log.Println(err)
	}
	f, err := p.Lookup("Print")
	if err != nil {
		log.Println(err)
	}
	f.(func())()
}

```
then I run main.

### What did you expect to see?
main process run and print "plugin"

### What did you see instead?
```
./main
SIGTRAP: trace trap
PC=0x7fb9410a81c1 m=0 sigcode=128

goroutine 0 [idle]:
runtime: unknown pc 0x7fb9410a81c1
stack: frame={sp:0x7ffe9713c240, fp:0x0} stack=[0x7ffe9693d8a0,0x7ffe9713c8e0)
00007ffe9713c140:  00007fb9410a64e0  00007ffe9713c230 
00007ffe9713c150:  0000000000000001  00007fb94383c548 
00007ffe9713c160:  0000000000000006  00000000000047a8 
00007ffe9713c170:  0000000000000ab0  0000000000000000 
00007ffe9713c180:  0000000000001000  0000000064cd305f 
00007ffe9713c190:  000056530156bc08  00007ffe9713c358 
00007ffe9713c1a0:  00007fb9410a68cc  00007ffe9713c230 
00007ffe9713c1b0:  00007ffe9713c240  00007fb9438491e9 
00007ffe9713c1c0:  0000000000000001  0000000000000000 
00007ffe9713c1d0:  0000000000000009  0000000000000000 
00007ffe9713c1e0:  0000000000000001  000056530156b8a0 
00007ffe9713c1f0:  0000000000000004  0000000000000000 
00007ffe9713c200:  000056530156b8a0  000056530156bc08 
00007ffe9713c210:  0000000000000000  0000000000000000 
00007ffe9713c220:  0000000000000000  0000000000000001 
00007ffe9713c230:  00000000ffffffff  00000000000037a8 
00007ffe9713c240: <00007fb9410a81ec  00007fb94384eb8a 
00007ffe9713c250:  0000000000000000  0000000000000001 
00007ffe9713c260:  00007ffe9713c9f8  00007ffe9713ca08 
00007ffe9713c270:  000056530156b8a0  00007fb94384ec91 
00007ffe9713c280:  78f25a33371354c3  00007ffe9713c5e0 
00007ffe9713c290:  00007ffe9713c480  fffffffffffffff0 
00007ffe9713c2a0:  0000000000000001  0000000000000006 
00007ffe9713c2b0:  000056530156b8a0  00007fb941013915 
00007ffe9713c2c0:  0000000000000000  00007ffe9713c4c0 
00007ffe9713c2d0:  00007fb9438522c0  00007ffe9713c430 
00007ffe9713c2e0:  4c576d5352587769  0000000000000000 
00007ffe9713c2f0:  0000000000000002  0000003200000008 
00007ffe9713c300:  0000000000000000  0000000000000000 
00007ffe9713c310:  0000000000000000  000000770000007c 
00007ffe9713c320:  0000005b0000006e  0000000000000000 
00007ffe9713c330:  00007fb9410a6000  0000000000000006 
runtime: unknown pc 0x7fb9410a81c1
stack: frame={sp:0x7ffe9713c240, fp:0x0} stack=[0x7ffe9693d8a0,0x7ffe9713c8e0)
00007ffe9713c140:  00007fb9410a64e0  00007ffe9713c230 
00007ffe9713c150:  0000000000000001  00007fb94383c548 
00007ffe9713c160:  0000000000000006  00000000000047a8 
00007ffe9713c170:  0000000000000ab0  0000000000000000 
00007ffe9713c180:  0000000000001000  0000000064cd305f 
00007ffe9713c190:  000056530156bc08  00007ffe9713c358 
00007ffe9713c1a0:  00007fb9410a68cc  00007ffe9713c230 
00007ffe9713c1b0:  00007ffe9713c240  00007fb9438491e9 
00007ffe9713c1c0:  0000000000000001  0000000000000000 
00007ffe9713c1d0:  0000000000000009  0000000000000000 
00007ffe9713c1e0:  0000000000000001  000056530156b8a0 
00007ffe9713c1f0:  0000000000000004  0000000000000000 
00007ffe9713c200:  000056530156b8a0  000056530156bc08 
00007ffe9713c210:  0000000000000000  0000000000000000 
00007ffe9713c220:  0000000000000000  0000000000000001 
00007ffe9713c230:  00000000ffffffff  00000000000037a8 
00007ffe9713c240: <00007fb9410a81ec  00007fb94384eb8a 
00007ffe9713c250:  0000000000000000  0000000000000001 
00007ffe9713c260:  00007ffe9713c9f8  00007ffe9713ca08 
00007ffe9713c270:  000056530156b8a0  00007fb94384ec91 
00007ffe9713c280:  78f25a33371354c3  00007ffe9713c5e0 
00007ffe9713c290:  00007ffe9713c480  fffffffffffffff0 
00007ffe9713c2a0:  0000000000000001  0000000000000006 
00007ffe9713c2b0:  000056530156b8a0  00007fb941013915 
00007ffe9713c2c0:  0000000000000000  00007ffe9713c4c0 
00007ffe9713c2d0:  00007fb9438522c0  00007ffe9713c430 
00007ffe9713c2e0:  4c576d5352587769  0000000000000000 
00007ffe9713c2f0:  0000000000000002  0000003200000008 
00007ffe9713c300:  0000000000000000  0000000000000000 
00007ffe9713c310:  0000000000000000  000000770000007c 
00007ffe9713c320:  0000005b0000006e  0000000000000000 
00007ffe9713c330:  00007fb9410a6000  0000000000000006 

goroutine 1 [syscall]:
runtime.cgocall(0x7fb9423d65f0, 0xc000362c40, 0xc000432018)
	/mnt/d/wsl/01.Tools/golang/1.16.3/go/src/runtime/cgocall.go:154 +0x65 fp=0xc000362c10 sp=0xc000362bd8 pc=0x7fb941c292c5
plugin._Cfunc_pluginOpen(0xc000438000, 0xc000432018, 0x0)
	_cgo_gotypes.go:76 +0x4e fp=0xc000362c40 sp=0xc000362c10 pc=0x7fb9423b386e
plugin.open.func1(0xc000438000, 0x1001, 0x1001, 0xc000432018, 0x7fb943832ae0)
	/mnt/d/wsl/01.Tools/golang/1.16.3/go/src/plugin/plugin_dlopen.go:64 +0x7d fp=0xc000362c70 sp=0xc000362c40 pc=0x7fb9423b4c3d
plugin.open(0x5653005ca015, 0xb, 0xc000000300, 0x300000002, 0xc000000300)
	/mnt/d/wsl/01.Tools/golang/1.16.3/go/src/plugin/plugin_dlopen.go:64 +0x34e fp=0xc000362ee8 sp=0xc000362c70 pc=0x7fb9423b3cee
plugin.Open(...)
	/mnt/d/wsl/01.Tools/golang/1.16.3/go/src/plugin/plugin.go:32
()
	/mnt/d/wsl/02.Code/01.GOPATH/src/plugindemo/main/main.go:9 +0x4a fp=0xc000362f88 sp=0xc000362ee8 pc=0x5653005c922a
runtime.main()
	/mnt/d/wsl/01.Tools/golang/1.16.3/go/src/runtime/proc.go:225 +0x2aa fp=0xc000362fe0 sp=0xc000362f88 pc=0x7fb941c72aca
runtime.goexit()
	/mnt/d/wsl/01.Tools/golang/1.16.3/go/src/runtime/asm_amd64.s:1371 +0x1 fp=0xc000362fe8 sp=0xc000362fe0 pc=0x7fb941cb3061

rax    0x0
rbx    0x7fb9410aa7b0
rcx    0x7ffe9713ca08
rdx    0x7ffe9713ca08
rdi    0x7fb9410ab040
rsi    0x7ffe9713c9f8
rbp    0x1
rsp    0x7ffe9713c240
r8     0x0
r9     0x1
r10    0x0
r11    0x7fb94109bbe0
r12    0x7ffe9713c9f8
r13    0x7ffe9713ca08
r14    0x7fb9410aa7b8
r15    0x0
rip    0x7fb9410a81c1
rflags 0x202
cs     0x33
fs     0x0
gs     0x0

```

then I use `objdump -D plugin.so` , I found that the function runtime.addmoduledata is not in plt of plugin.so, when go.link.addmoduledata is called in dllopen,  it jumped to a wrong address.
![image](https://user-images.githubusercontent.com/34932896/115690990-34b7a200-a390-11eb-983c-84fbc04fb711.png)

If I can add `runtime.addmoduledata` to plt in plugin.so, maybe I can use plugin.Open to open the plugin.so success.so,I modified the linker in toolchain. then this mistake is fixed. 

![image](https://user-images.githubusercontent.com/34932896/115693316-63cf1300-a392-11eb-9e16-8d331fa38729.png)
![image](https://user-images.githubusercontent.com/34932896/115694242-42225b80-a393-11eb-8a51-3fe5fd7f5be8.png)

Then, when I run the main. it returns an error:  "plugin was built with a different version of package". I found the pkghash of the pkgs in libstd.so is blank in plugin. I modifier the linker:

![image](https://user-images.githubusercontent.com/34932896/115693652-b4467080-a392-11eb-81dd-327ff22c5938.png)

It will not generate pkghash of the pkg in libstd.so to plugin.so in linktime.

After these modification，the main process runs success，and I tried some more complex plugin，it works well.
But I can't  judge are there any other risk after these modification. I want to use it in production environment. Can anyone help me?

--

**Comment by crazycomputer** [comment]
[https://go-review.googlesource.com/c/go/+/61090](url)

**Comment by crazycomputer** [comment]
@ianlancetaylor

**Comment by crazycomputer** [comment]
#18671

**Comment by cherrymui** [comment]
There is currently no plan to support plugin with linkshared.
You already found #18671 . Closing this as a dup. Thanks.

**Comment by ChenJhua** [comment]
@crazycomputer Do you get an error using plugin.so after you change it? If not I'd like to refer to your changes as I also need this functionality

**Comment by zhouguangyuan0718** [comment]
> @crazycomputer Do you get an error using plugin.so after you change it? If not I'd like to refer to your changes as I also need this functionality

@ChenJhua 
This is my new account in github instead of crazycomputer.
The change in this issue is based on go 1.14 and there is some bugs in it.
I had developed another two new patches for the functionality based on 1.16 and it can work on master. 

The one patch is compatible with official plugin, but there is a huge modification in linker.
In the other patch, I had modified a little in official plugin library in runtime and less modification in linker. I am planing to use this patch in our producting environment. 

If you can use the second one, I will try to shared it to you or open source of it.
If you can only use a patch which is compatible with official plugin (the patch one). I'm not sure I can shared the patch to you before investigation. Because it is really huge and maybe it is risky.

**Comment by ChenJhua** [comment]
@zhouguangyuan0718 I would like to use the second plan, share it with me if you can, let's contact by email or QQ. My email account is: 370619197@qq.com and QQ is the same. Thank you very much.

**Comment by ChenJhua** [comment]
@zhouguangyuan0718 If you don't mind, you can also share the code here. Thank you very much.

**Comment by zhouguangyuan0718** [comment]
> @zhouguangyuan0718 If you don't mind, you can also share the code here. Thank you very much.

Communicating on github is better. I will send a CL to shared the code with you tonight. Maybe 8 hours ago.
By the way,  the change in this issue can work in some situations. What error do you get with the change?

**Comment by ChenJhua** [comment]
I modified the code without recompiling golang. and then re-executed it `go build -buildmode=plugin -linkshared myplugin2.go`.it is ok. but i run main.go.The following error is reported
```
SIS3.0.60.0 /data/cjh/go/goplugin # go run main.go 
SIGTRAP: trace trap
PC=0x7f97718d2741 m=0 sigcode=128

goroutine 0 [idle]:
runtime: unknown pc 0x7f97718d2741
stack: frame={sp:0x7fff049f3120, fp:0x0} stack=[0x7fff041f4638,0x7fff049f3670)
00007fff049f3020:  00007f97718d1480  00007f9771ad3fa0 
00007fff049f3030:  00007fff049f3170  00007f979ade40cf 
00007fff049f3040:  00007f9700000001  0000000000000000 
00007fff049f3050:  0000001f0000007c  00007f979afe9000 
00007fff049f3060:  00007fff049f30e0  00007fff049f30e8 
00007fff049f3070:  00007fff049f30f8  0000000000000000 
00007fff049f3080:  00007fff049f3158  0000000000000000 
00007fff049f3090:  0000000000000000  00007fff049f30f0 
00007fff049f30a0:  00007f97718d1720  00007fff049f3158 
00007fff049f30b0:  00007f97718d1000  000000000240dd88 
00007fff049f30c0:  00007f97718d1b4e  0000000000000000 
00007fff049f30d0:  00007f97718d2518  00007f97718d12d0 
00007fff049f30e0:  000000000240dff0  00007f979ade915e 
00007fff049f30f0:  0000000000429450  0000000000000005 
00007fff049f3100:  00007f97718d1bd0  0000000000000948 
00007fff049f3110:  000000000000003d  0000000000000000 
00007fff049f3120: <00007f97718d276c  00007f979ade79b3 
00007fff049f3130:  000000000240da30  0000000000000006 
00007fff049f3140:  8000000000000000  0000000000000000 
00007fff049f3150:  00007fff049f3230  0000000000000006 
00007fff049f3160:  0000000000000001  0000000000000006 
00007fff049f3170:  000000000240da30  00007f979adec58e 
00007fff049f3180:  000000000240da30  000000000240dff0 
00007fff049f3190:  00007fff049f3230  00007f979adec349 
00007fff049f31a0:  0000000000000000  0000000000000000 
00007fff049f31b0:  0000000000000001  000000000240da30 
00007fff049f31c0:  00007fff049f3180  00007fff049f31a0 
00007fff049f31d0:  0000000000000001  0000000000000006 
00007fff049f31e0:  00007f979affb120  00007fff049f3170 
00007fff049f31f0:  8000010200000001  00007fff049f3398 
00007fff049f3200:  0000000000484b2e <runtime.main+622>  00007f979afe6728 
00007fff049f3210:  0000000080000102  000000c00008a000 
runtime: unknown pc 0x7f97718d2741
stack: frame={sp:0x7fff049f3120, fp:0x0} stack=[0x7fff041f4638,0x7fff049f3670)
00007fff049f3020:  00007f97718d1480  00007f9771ad3fa0 
00007fff049f3030:  00007fff049f3170  00007f979ade40cf 
00007fff049f3040:  00007f9700000001  0000000000000000 
00007fff049f3050:  0000001f0000007c  00007f979afe9000 
00007fff049f3060:  00007fff049f30e0  00007fff049f30e8 
00007fff049f3070:  00007fff049f30f8  0000000000000000 
00007fff049f3080:  00007fff049f3158  0000000000000000 
00007fff049f3090:  0000000000000000  00007fff049f30f0 
00007fff049f30a0:  00007f97718d1720  00007fff049f3158 
00007fff049f30b0:  00007f97718d1000  000000000240dd88 
00007fff049f30c0:  00007f97718d1b4e  0000000000000000 
00007fff049f30d0:  00007f97718d2518  00007f97718d12d0 
00007fff049f30e0:  000000000240dff0  00007f979ade915e 
00007fff049f30f0:  0000000000429450  0000000000000005 
00007fff049f3100:  00007f97718d1bd0  0000000000000948 
00007fff049f3110:  000000000000003d  0000000000000000 
00007fff049f3120: <00007f97718d276c  00007f979ade79b3 
00007fff049f3130:  000000000240da30  0000000000000006 
00007fff049f3140:  8000000000000000  0000000000000000 
00007fff049f3150:  00007fff049f3230  0000000000000006 
00007fff049f3160:  0000000000000001  0000000000000006 
00007fff049f3170:  000000000240da30  00007f979adec58e 
00007fff049f3180:  000000000240da30  000000000240dff0 
00007fff049f3190:  00007fff049f3230  00007f979adec349 
00007fff049f31a0:  0000000000000000  0000000000000000 
00007fff049f31b0:  0000000000000001  000000000240da30 
00007fff049f31c0:  00007fff049f3180  00007fff049f31a0 
00007fff049f31d0:  0000000000000001  0000000000000006 
00007fff049f31e0:  00007f979affb120  00007fff049f3170 
00007fff049f31f0:  8000010200000001  00007fff049f3398 
00007fff049f3200:  0000000000484b2e <runtime.main+622>  00007f979afe6728 
00007fff049f3210:  0000000080000102  000000c00008a000 

goroutine 1 [syscall]:
runtime.cgocall(0x51ad90, 0xc00004bc40, 0xc000010028)
	/usr/local/go/src/runtime/cgocall.go:154 +0x5b fp=0xc00004bc10 sp=0xc00004bbd8 pc=0x450a5b
plugin._Cfunc_pluginOpen(0xc00008a000, 0xc000010028, 0x0)
	_cgo_gotypes.go:76 +0x45 fp=0xc00004bc40 sp=0xc00004bc10 pc=0x519765
plugin.open.func1(0xc00008a000, 0x1001, 0x1001, 0xc000010028, 0x82acc0)
	/usr/local/go/src/plugin/plugin_dlopen.go:64 +0x71 fp=0xc00004bc70 sp=0xc00004bc40 pc=0x51a911
plugin.open(0x547ae4, 0xc, 0xc000000180, 0x300000002, 0xc000000180)
	/usr/local/go/src/plugin/plugin_dlopen.go:64 +0x345 fp=0xc00004bee8 sp=0xc00004bc70 pc=0x519ba5
plugin.Open(...)
	/usr/local/go/src/plugin/plugin.go:32
main.main()
	/data/cjh/go/goplugin/main.go:6 +0x48 fp=0xc00004bf88 sp=0xc00004bee8 pc=0x51ab08
runtime.main()
	/usr/local/go/src/runtime/proc.go:225 +0x256 fp=0xc00004bfe0 sp=0xc00004bf88 pc=0x484b16
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:1371 +0x1 fp=0xc00004bfe8 sp=0xc00004bfe0 pc=0x4b54a1

rax    0x0
rbx    0x1
rcx    0x7f9771ad3dc0
rdx    0x7fff049f3788
rdi    0x7f9771ad4020
rsi    0x7fff049f3778
rbp    0x7fff049f3778
rsp    0x7fff049f3120
r8     0x7f979affb150
r9     0x7f979affb150
r10    0x0
r11    0x206
r12    0x7fff049f3788
r13    0x7f9771ad3750
r14    0x0
r15    0x7f9771ad3748
rip    0x7f97718d2741
rflags 0x283
cs     0x33
fs     0x0
gs     0x0
exit status 2
```
Now I'm going to recompile golang to try.
```
$ /data/cjh/go/goplugin # cat myplugin2.go 
package main
import "fmt"
func Hello() {
    fmt.Println("Hello, World! I am plugin 2")
}
$ /data/cjh/go/goplugin # cat main.go 
package main
import "log"
import "plugin"
func main() {
	//	加载 myplugin 库
	p, err := plugin.Open("myplugin2.so")
	if err != nil {
		log.Fatal(err)
	}
	//	取得 Hello 函数
	fn, err := p.Lookup("Hello")
	if err != nil {
		log.Fatal(err)
	}
	//	调用函数
        fn.(func())()
}

```

**Comment by zhouguangyuan0718** [comment]
The cmd should be rebuild, because the linker is modified. 
This error is easy to resolve.  For go 1.14, you can resolve by the patch in the picture of this issue.
For go 1.16, 1.17 and tips, I will send a CL tonight.

**Comment by ChenJhua** [comment]
> The cmd should be rebuild, because the linker is modified. This error is easy to resolve. For go 1.14, you can resolve by the patch in the picture of this issue. For go 1.16, 1.17 and tips, I will send a CL tonight.

I am currently on go 1.16. Also trying the changes in your picture to recompile and test. Looking forward to the CL you shared

**Comment by zhouguangyuan0718** [comment]
> > The cmd should be rebuild, because the linker is modified. This error is easy to resolve. For go 1.14, you can resolve by the patch in the picture of this issue. For go 1.16, 1.17 and tips, I will send a CL tonight.
> 
> I am currently on go 1.16. Also trying the changes in your picture to recompile and test. Looking forward to the CL you shared

I sent a CL  https://go-review.googlesource.com/c/go/+/369594 for  the second patch.
It's a little different from the patch for our production environment. But is should work, too. If there is any problem, please ping me.

**Comment by ChenJhua** [comment]
> > > The cmd should be rebuild, because the linker is modified. This error is easy to resolve. For go 1.14, you can resolve by the patch in the picture of this issue. For go 1.16, 1.17 and tips, I will send a CL tonight.
> > 
> > 
> > I am currently on go 1.16. Also trying the changes in your picture to recompile and test. Looking forward to the CL you shared
> 
> I sent a CL https://go-review.googlesource.com/c/go/+/369594 for the second patch. It's a little different from the patch for our production environment. But is should work, too. If there is any problem, please ping me.

Thank you very much for providing me the code so late, I will try it tomorrow. thank you very much

**Comment by zhouguangyuan0718** [comment]
@ChenJhua 
For go 1.16, you should edit the file src/cmd/compile/internal/gc/reflect.go instead of src/cmd/compile/internal/reflectdata/reflect.go.
Please use this patch instead of CL for 1.16.

<pre>

diff --git a/src/cmd/compile/internal/gc/reflect.go b/src/cmd/compile/internal/gc/reflect.go
index 9401eba7a5..7cc4715814 100644
--- a/src/cmd/compile/internal/gc/reflect.go
+++ b/src/cmd/compile/internal/gc/reflect.go
@@ -1587,13 +1587,13 @@ func dumptabs() {
                        // Dump ptab symbol into go.pluginsym package.
                        //
                        // type ptab struct {
-                       //      name nameOff
-                       //      typ  typeOff // pointer to symbol
+                       //       name name
+                       //       typ  *_type
                        // }
                        nsym := dname(p.s.Name, "", nil, true)
                        tsym := dtypesym(p.t)
-                       ot = dsymptrOff(s, ot, nsym)
-                       ot = dsymptrOff(s, ot, tsym)
+                       ot = dsymptr(s, ot, nsym, 0)
+                       ot = dsymptr(s, ot, tsym, 0)
                        // Plugin exports symbols as interfaces. Mark their types
                        // as UsedInIface.
                        tsym.Set(obj.AttrUsedInIface, true)
diff --git a/src/cmd/link/internal/ld/ld.go b/src/cmd/link/internal/ld/ld.go
index 7ff9c41f96..2bfdf826f8 100644
--- a/src/cmd/link/internal/ld/ld.go
+++ b/src/cmd/link/internal/ld/ld.go
@@ -253,8 +253,12 @@ func PrepareAddmoduledata(ctxt *Link) (*loader.SymbolBuilder, loader.Sym) {
        ctxt.loader.SetAttrLocal(ifs, true)
        initfunc.SetType(sym.STEXT)

+       if ctxt.BuildMode == BuildModePlugin && ctxt.linkShared {
+               ambd := ctxt.loader.MakeSymbolUpdater(amd)
+               ambd.SetType(sym.SDYNIMPORT)
+       }
        // Add the init func and/or addmoduledata to Textp.
-       if ctxt.BuildMode == BuildModePlugin {
+       if ctxt.BuildMode == BuildModePlugin && !ctxt.linkShared {
                ctxt.Textp = append(ctxt.Textp, amd)
        }
        ctxt.Textp = append(ctxt.Textp, initfunc.Sym())
diff --git a/src/cmd/link/internal/ld/lib.go b/src/cmd/link/internal/ld/lib.go
index 18db567041..bae5d2d12b 100644
--- a/src/cmd/link/internal/ld/lib.go
+++ b/src/cmd/link/internal/ld/lib.go
@@ -536,7 +536,7 @@ func (ctxt *Link) loadlib() {

        // Plugins a require cgo support to function. Similarly, plugins may require additional
        // internal linker support on some platforms which may not be implemented.
-       ctxt.canUsePlugins = ctxt.LibraryByPkg["plugin"] != nil && iscgo
+       ctxt.canUsePlugins = ctxt.LibraryByPkg["plugin"] != nil && iscgo || ctxt.PackageShlib["plugin"] != ""

        // We now have enough information to determine the link mode.
        determineLinkMode(ctxt)
diff --git a/src/cmd/link/internal/ld/symtab.go b/src/cmd/link/internal/ld/symtab.go
index f54cf9ea2f..492d0e53fc 100644
--- a/src/cmd/link/internal/ld/symtab.go
+++ b/src/cmd/link/internal/ld/symtab.go
@@ -708,7 +708,7 @@ func (ctxt *Link) symtab(pcln *pclntab) []sym.SymKind {
                if ldr.SymType(ptab) != sym.SRODATA {
                        panic(fmt.Sprintf("go.plugin.tabs is %v, not SRODATA", ldr.SymType(ptab)))
                }
-               nentries := uint64(len(ldr.Data(ptab)) / 8) // sizeof(nameOff) + sizeof(typeOff)
+               nentries := uint64(len(ldr.Data(ptab)) / (2 * ctxt.Arch.PtrSize)) // sizeof(nameOff) + sizeof(typeOff)
                moduledata.AddAddr(ctxt.Arch, ptab)
                moduledata.AddUint(ctxt.Arch, nentries)
                moduledata.AddUint(ctxt.Arch, nentries)
@@ -723,8 +723,12 @@ func (ctxt *Link) symtab(pcln *pclntab) []sym.SymKind {
                pkghashes := ldr.CreateSymForUpdate("go.link.pkghashes", 0)
                pkghashes.SetLocal(true)
                pkghashes.SetType(sym.SRODATA)
-
+               var shlibs uint64
                for i, l := range ctxt.Library {
+                       if ctxt.linkShared && l.Shlib != "" {
+                               shlibs++
+                               continue
+                       }
                        // pkghashes[i].name
                        addgostring(ctxt, ldr, pkghashes, fmt.Sprintf("go.link.pkgname.%d", i), l.Pkg)
                        // pkghashes[i].linktimehash
@@ -734,8 +738,8 @@ func (ctxt *Link) symtab(pcln *pclntab) []sym.SymKind {
                        pkghashes.AddAddr(ctxt.Arch, hash)
                }
                moduledata.AddAddr(ctxt.Arch, pkghashes.Sym())
-               moduledata.AddUint(ctxt.Arch, uint64(len(ctxt.Library)))
-               moduledata.AddUint(ctxt.Arch, uint64(len(ctxt.Library)))
+               moduledata.AddUint(ctxt.Arch, uint64(len(ctxt.Library))-shlibs)
+               moduledata.AddUint(ctxt.Arch, uint64(len(ctxt.Library))-shlibs)
        } else {
                moduledata.AddUint(ctxt.Arch, 0) // pluginpath
                moduledata.AddUint(ctxt.Arch, 0)
diff --git a/src/runtime/plugin.go b/src/runtime/plugin.go
index 5e05be71ec..72fae1a647 100644
--- a/src/runtime/plugin.go
+++ b/src/runtime/plugin.go
@@ -78,8 +78,7 @@ func plugin_lastmoduleinit() (path string, syms map[string]interface{}, errstr s
        // a dependency on the reflect package.
        syms = make(map[string]interface{}, len(md.ptab))
        for _, ptab := range md.ptab {
-               symName := resolveNameOff(unsafe.Pointer(md.types), ptab.name)
-               t := (*_type)(unsafe.Pointer(md.types)).typeOff(ptab.typ)
+               symName, t := ptab.name, ptab.typ
                var val interface{}
                valp := (*[2]unsafe.Pointer)(unsafe.Pointer(&val))
                (*valp)[0] = unsafe.Pointer(t)
@@ -131,6 +130,6 @@ func inRange(r0, r1, v0, v1 uintptr) bool {
 // and global variable in the main package of a plugin. It is used to
 // initialize the plugin module's symbol map.
 type ptabEntry struct {
-       name nameOff
-       typ  typeOff
+       name name
+       typ  *_type
 }
</pre>

**Comment by ChenJhua** [comment]
> go build -linkshared main.go

yes，it is good!
these changes can be adapted to my environment. Thank you very much. 
I would also like to ask you how you analyzed the problem and came up with the above changes. 
I would like to learn the whole process to improve my own capabilities. 
If possible, would you be willing to share the whole process with me?

**Comment by zhouguangyuan0718** [comment]
At the beginning,  I got a fatal "trace trap",too. By disassembly the plugin biniary, I found the plugin miss function "addmoduledata". So the first thing to do is to fix the missing function. I read the code of linker for a long time. Then I understood the how the symbol addmoduledata generated. So I got the first point to fix.

After the first fix, I found that the plugin don't crash when call dlopen for it. But it crashed with "plugin was built with a different version of package " for std library. Then I found out how the pkghash generated. And skip generating pkghash for shared libraries.

At this time, the changes can work in all the situations I need.  So, I opened this issue, but unfortunately this function is not on the roadmap of offical. 

Then I run the testcases of offical plugin to test plugin with linkshared. It crashed. It is because of the reloction type of ptabs. So I changed it, and it seems run well.

I will try to follow the offical shared support as long as I can. But I really don't know how long I can follow. Maybe the linkshared mode will be removed or changed a lot in the future. https://github.com/golang/go/issues/47788
